ddl/ddl_db_change_test.go

```diff
@@ -628,3 +631,97 @@ func (s *testStateChangeSuite) TestCreateDBIfNotExists(c *C) {
 	defer s.se.Execute(context.Background(), "drop database test_not_exists")
 	s.testParallelExecSQL(c, "create database if not exists test_not_exists;")
 }
+
+// TestParallelDDLBeforeRunDDLJob tests a session to execute DDL with an outdated information schema.
+func (s *testStateChangeSuite) TestParallelDDLBeforeRunDDLJob(c *C) {
+	defer s.se.Execute(context.Background(), "drop table test_table")
+	_, err := s.se.Execute(context.Background(), "use test_db_state")
+	c.Assert(err, IsNil)
+	_, err = s.se.Execute(context.Background(), "create table test_table (c1 int, c2 int default 1, index (c1))")
+	c.Assert(err, IsNil)
+
+	// Create two sessions.
+	se, err := session.CreateSession(s.store)
+	c.Assert(err, IsNil)
+	_, err = se.Execute(context.Background(), "use test_db_state")
+	c.Assert(err, IsNil)
+	se1, err := session.CreateSession(s.store)
+	c.Assert(err, IsNil)
+	_, err = se1.Execute(context.Background(), "use test_db_state")
+	c.Assert(err, IsNil)
+
+	ddl.IsInTest = true
+	defer func() { ddl.IsInTest = false }()
+	callback := &ddl.TestDDLCallback{}
+	var (
+		testSessionCnt        int32
+		testObtainedSchemaCnt int32 // testObtainedSchemaCnt is the number of sessions that have obtained information schema.
+		testFirstConnID       = uint64(math.MaxUint64)
+	)
+	callback.OnGetInfoSchemaExported = func(ctx sessionctx.Context, fn ddl.GetInfoSchema) infoschema.InfoSchema {
+		// The following code is for testing.
+		// Make srue the two sessions get the same information schema before executing DDL.
+		// After the first session executes its DDL, then the second session executes its DDL.
+		var info infoschema.InfoSchema
+		interval := 5 * time.Millisecond
+		atomic.AddInt32(&testSessionCnt, 1)
+		for {
+			cnt := atomic.LoadInt32(&testSessionCnt)
+			// Make sure there are two sessions running here.
+			if cnt == 2 {
+				info = fn()
+				atomic.AddInt32(&testObtainedSchemaCnt, 1)
+				break
+			}
+			time.Sleep(interval)
+		}
+		currID := ctx.GetSessionVars().ConnectionID
+		if currID == 1 {
+			atomic.CompareAndSwapUint64(&testFirstConnID, math.MaxUint64, currID)
+			time.Sleep(interval)
+		}
+		for {
+			firstID := atomic.LoadUint64(&testFirstConnID)
+			seCnt := atomic.LoadInt32(&testSessionCnt)
+			obtainedSchemaCnt := atomic.LoadInt32(&testObtainedSchemaCnt)
+			// Make sure the two session have got the same information schema. And the first session can continue to go on,
+			// or the frist session finished this SQL, then other sessions can continue to go on.
+			if obtainedSchemaCnt == 2 && (firstID == currID || seCnt == 0) {
+				break
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+
+		return info
+	}
+	d := s.dom.DDL()
+	d.SetHook(callback)
+
+	// Make sure we
+	ch := make(chan struct{})
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+
+		close(ch)
```

Comment:
Why need a channel here?

Reply:
Hope to run this goroutine before running the next goroutine.